pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java

```diff
@@ -91,6 +95,18 @@ public void asyncReadEntriesOrWait(ManagedCursor cursor, int numberOfEntriesToRe
                 compactedTopicContext.thenCompose(
                     (context) -> findStartPoint(cursorPosition, context.ledger.getLastAddConfirmed(), context.cache)
                         .thenCompose((startPoint) -> {
+                            // do not need to read the compaction ledger if it is empty.
+                            // the cursor just needs to be set to the compaction horizon
+                            if (startPoint == COMPACT_LEDGER_EMPTY) {
+                                try {
+                                    MessageIdImpl msgId =
+                                      (MessageIdImpl) LedgerMetadataUtils.getCompactedToMessageId(context.ledger);
+                                    log.info("compaction horizaion msg id: {}", msgId);
+                                    cursor.seek(PositionImpl.get(msgId.getLedgerId(), msgId.getEntryId()));
+                                } catch (IOException e) {
+                                    new RuntimeException(e);
+                                }
+                            }
```

Comment:
`else if (..` ?

Reply:
we can just return CompletableFuture.completedFuture(null);